src/lib/network_pool/test/indexed_pool_tests.ml

```diff
@@ -830,38 +834,89 @@ let rec interleave_at_random queues =
+  Quickcheck.test gen_accounts_and_transactions ~f:(fun (account_map, txns) ->
       let pool = pool_of_transactions ~account_map txns |> rem_lowest_fee 5 in
       let txns = Sequence.to_list @@ Indexed_pool.transactions ~logger pool in
       with_accounts txns ~init:() ~account_map ~f:(fun () a t ->
           [%test_eq: Account_nonce.t] (txn_nonce t) a.nonce |> Result.return )
       |> Result.ok_or_failwith ;
       assert_invariants pool )
 
-let construct_ledger accounts =
-  Mina_ledger.Ledger.with_ledger ~depth:4 ~f:(fun ledger ->
-      List.iter accounts ~f:(fun a ->
-          Mina_ledger.Ledger.create_new_account_exn ledger
-            (Account.identifier a) a ) ;
-      ledger )
-
-(* let revalidation_drops_nothing_unless_ledger_changed () = *)
-(*   Quickcheck.test *)
-(*     (let open Quickcheck.Generator.Let_syntax in) *)
-(*     ~f:(fun () -> ()) *)
+let revalidation_drops_nothing_unless_ledger_changed () =
+  Quickcheck.test gen_accounts_and_transactions ~f:(fun (account_map, txns) ->
+      let pool = pool_of_transactions ~account_map txns in
+      let pool', dropped =
+        Indexed_pool.revalidate pool ~logger `Entire_pool (fun aid ->
+            Public_key.Compressed.Map.find_exn account_map
+              (Account_id.public_key aid) )
+      in
+      [%test_eq:
+        Transaction_hash.User_command_with_valid_signature.t Sequence.t] dropped
+        Sequence.empty ;
+      [%test_eq: Indexed_pool.t] pool pool' ;
+      let to_apply = Indexed_pool.transactions ~logger pool in
+      let to_apply' = Indexed_pool.transactions ~logger pool' in
+      [%test_eq:
+        Transaction_hash.User_command_with_valid_signature.t Sequence.t]
+        to_apply to_apply' )
+
+let apply_transactions txns accounts =
+  List.fold txns ~init:accounts ~f:(fun m t ->
+      let txn = Signed_command.forget_check t in
+      let sender = Public_key.compress txn.signer in
+      Public_key.Compressed.Map.update m sender ~f:(function
+        | None ->
+            failwith "sender not found"
+        | Some a ->
+            let open Account.Poly in
+            let nonce = Account.Nonce.succ a.nonce in
+            let balance =
+              let amt =
+                Signed_command.amount txn |> Option.value ~default:Amount.zero
+              in
+              Balance.(sub_amount a.balance amt)
+              |> Option.value ~default:Balance.zero
+            in
+            { a with nonce; balance } ) )
+
+let txn_hash = Transaction_hash.User_command_with_valid_signature.hash
+
+let application_invalidates_applied_transactions () =
+  Quickcheck.test
+    (let open Quickcheck.Generator.Let_syntax in
+    let%bind accounts, txns = gen_accounts_and_transactions in
+    (* Simulate application of some transactions in order to invalidate them. *)
+    let%map app_count = Int.gen_incl 0 (List.length txns) in
+    let updated_accounts =
+      apply_transactions (List.take txns app_count) accounts
```

Comment:
Do I understand right that you try to invalidate only a prefix of transcations? Maybe it's worth exploring invalidating some transactions that are not prefix (though nonces need to be taken care of)? Not sure if generality suffers overall with the current scheme, but this is a suspicion I have.

Reply:
This test is specifically about invalidating by application. You can only apply the prefix of the queue, otherwise you'll get a failure (in real life at least). I thought about invalidating other transactions, but this could only happen either due to balance decrease or nonce increase, both of which happen only when a transaction gets applied, so I figure this is sufficient. Theoretically I can imagine a situation, when a user generates competing transactions with the same nonce, sending them to different nodes and it so happens that the block producer gets only one of them and applies it – but that also always happens at the head of the queue.